internal/appsec/trace/tagsholder.go

```diff
@@ -5,29 +5,64 @@
 
 package trace
 
-import "sync"
+import (
+	"encoding/json"
+	"sync"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type serializableTag struct {
+	tag any
+}
+
+func (t *serializableTag) MarshalJSON() ([]byte, error) {
+	return json.Marshal(t.tag)
+}
 
 // TagsHolder wraps a map holding tags. The purpose of this struct is to be
 // used by composition in an Operation to allow said operation to handle tags
 // addition/retrieval.
 type TagsHolder struct {
-	tags map[string]interface{}
-	mu   sync.Mutex
+	tags map[string]any
+	mu   sync.RWMutex
 }
 
 // NewTagsHolder returns a new instance of a TagsHolder struct.
 func NewTagsHolder() TagsHolder {
-	return TagsHolder{tags: map[string]interface{}{}}
+	return TagsHolder{tags: make(map[string]any)}
 }
 
 // AddTag adds the key/value pair to the tags map
-func (m *TagsHolder) AddTag(k string, v interface{}) {
+func (m *TagsHolder) AddTag(k string, v any) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	m.tags[k] = v
 }
 
-// Tags returns the tags map
-func (m *TagsHolder) Tags() map[string]interface{} {
-	return m.tags
+// AddSerializableTag adds the key/value pair to the tags map. Value is serialized
+func (m *TagsHolder) AddSerializableTag(k string, v any) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.tags[k] = &serializableTag{tag: v}
+}
+
+// Tags returns a copy of the aggregated tags map (normal and serialized)
+func (m *TagsHolder) Tags() map[string]any {
+	tags := make(map[string]any, len(m.tags))
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for k, v := range m.tags {
+		tags[k] = v
+		marshaler, ok := v.(json.Marshaler)
```

Comment:
I think you should limit the type-assert to `serializableTag` only to avoid serializing by accident things that were added with `AddTag`:
```suggestion
		marshaler, ok := v.(serializableTag)
```

Reply:
39b0728c5d6bc2251e17